File: scripts/sync_versions.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.join(__dirname, '..');

const pkgPath = path.join(ROOT, 'package.json');
const pubspecPath = path.join(ROOT, 'packages', 'steward_cli', 'pubspec.yaml');

if (!fs.existsSync(pkgPath)) {
  console.error(`package.json not found at ${pkgPath}`);
  process.exit(1);
}

if (!fs.existsSync(pubspecPath)) {
  console.error(`pubspec.yaml not found at ${pubspecPath}`);
  process.exit(1);
}

const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
const version = pkg.version;

if (!version) {
  console.error('No version found in package.json');
  process.exit(1);
}

let pubspec = fs.readFileSync(pubspecPath, 'utf8');
const versionRegex = /^version:\s*[^\r\n]+/m;

if (!versionRegex.test(pubspec)) {
  console.error('Could not find version line in pubspec.yaml');
  process.exit(1);
}

pubspec = pubspec.replace(versionRegex, `version: ${version}`);
fs.writeFileSync(pubspecPath, pubspec, 'utf8');

console.log(`Synced version: packages/steward_cli/pubspec.yaml is now at ${version}`);

File: packages/steward_cli/tool/sync_versions.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as p;

void main() {
  // Resolve the script location to find package roots relative to it
  final scriptPath = File(Platform.script.toFilePath()).canonicalize().path;
  final toolDir = Directory(p.dirname(scriptPath));
  final stewardCliDir = toolDir.parent;
  final rootDir = stewardCliDir.parent.parent;

  final pkgFile = File(p.join(rootDir.path, 'package.json'));
  final pubspecFile = File(p.join(stewardCliDir.path, 'pubspec.yaml'));

  if (!pkgFile.existsSync()) {
    stderr.writeln('Error: package.json not found at ${pkgFile.path}');
    exit(1);
  }

  if (!pubspecFile.existsSync()) {
    stderr.writeln('Error: pubspec.yaml not found at ${pubspecFile.path}');
    exit(1);
  }

  final pkgJson = jsonDecode(pkgFile.readAsStringSync()) as Map<String, dynamic>;
  final version = pkgJson['version'] as String?;

  if (version == null) {
    stderr.writeln('Error: No version found in package.json');
    exit(1);
  }

  var pubspec = pubspecFile.readAsStringSync();
  final versionRegex = RegExp(r'^version:\s*[^\r\n]+', multiLine: true);

  if (!versionRegex.hasMatch(pubspec)) {
    stderr.writeln('Error: Could not find version line in pubspec.yaml');
    exit(1);
  }

  pubspec = pubspec.replaceFirst(versionRegex, 'version: $version');
  pubspecFile.writeAsStringSync(pubspec);

  stdout.writeln('Synced version: packages/steward_cli/pubspec.yaml is now at $version');
}

extension on File {
  File canonicalize() => File(p.canonicalize(path));
}
